src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +311,146 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
+        // just scheduled, but it makes the code a lot simpler (compared to having a separate, synchronous code path).
+        // Cleverness: swallow any exceptions - they'll be surfaced by GetKeyRingFromCompletedTask, if appropriate.
+        existingTask
+            .ContinueWith(static _ => { }, TaskScheduler.Default)
+            .Wait();
+
+        var newKeyRing = GetKeyRingFromCompletedTask(existingTask, utcNow); // Throws if the task failed
+        if (newKeyRing is null)
+        {
+            // Another thread won - check whether it cached a new key ring
+            var newCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (newCacheableKeyRing is null)
+            {
+                // There will have been a better exception from the winning thread
+                throw Error.KeyRingProvider_RefreshFailedOnOtherThread();
+            }
+
+            newKeyRing = newCacheableKeyRing.KeyRing;
+        }
+
+        return newKeyRing;
+    }
+
+    /// <summary>
+    /// Returns null if another thread already processed the completed task.
+    /// Otherwise, if the given completed task completed successfully, clears the task
+    /// and either caches and returns the resulting key ring or throws, according to the
+    /// successfulness of the task.
+    /// </summary>
+    private IKeyRing? GetKeyRingFromCompletedTask(Task<CacheableKeyRing> task, DateTime utcNow)
+    {
+        Debug.Assert(task.IsCompleted);
+
+        lock (_cacheableKeyRingLockObj)
+        {
+            // If the parameter doesn't match the field, another thread has already consumed the task (and it's reflected in _cacheableKeyRing)
+            if (!ReferenceEquals(task, Volatile.Read(ref _cacheableKeyRingTask)))
+            {
+                return null;
+            }
+
+            Volatile.Write(ref _cacheableKeyRingTask, null);
+
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                var newCacheableKeyRing = task.Result;
+                Volatile.Write(ref _cacheableKeyRing, newCacheableKeyRing);
+
+                // An unconsumed task result is considered to satisfy forceRefresh.  One could quibble that this isn't really
+                // a forced refresh, but we'll still return a key ring newer than the one the caller was dissatisfied with.
+                return newCacheableKeyRing.KeyRing;
+            }
+
+            var exception = task.Exception!;
+
+            var existingCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (existingCacheableKeyRing is not null && !CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow))
+            {
+                // If reading failed, we probably don't want to try again for a little bit, so slightly extend the
+                // lifetime of the current cache entry
+                Volatile.Write(ref _cacheableKeyRing, existingCacheableKeyRing.WithTemporaryExtendedLifetime(utcNow));
```

Comment:
We should probably guard against extending the cache lifetime beyond the default key lifetime, but that's not new.

Reply:
https://github.com/dotnet/aspnetcore/issues/54708